### DISCUSSION TITLE
Changed behaviour on Mapper columns for polymorphic tables

### DISCUSSION THREAD

**Discussion by liverpool1026** [sqlalchemy/sqlalchemy#10211] [Usage Questions]
Hi, we have migrated from 1.3 -> 1.4 recently and is seeing this change.
I don't seem to find this documented in the change logs in 1.4 (or I just missed it).

### Code
Table Structure
```
class ORMBase(declarative_base()):
    __abstract__ = True

    @declared_attr
    def __tablename__(cls) -> str:
        return cls.__name__.lower()


class Group(ORMBase):
    __mapper_args__ = {
        "polymorphic_identity": "group",
        "polymorphic_on": "type",
    }
    type = Column(String, nullable=False)
    id = Column(Integer, primary_key=True, nullable=False)
    group_name = Column(String, nullable=False)


class ChildGroup(Group):
    id = Column(ForeignKey("group.id"), primary_key=True, nullable=False)
    __mapper_args__ = {
        "polymorphic_identity": "childgroup",
        "polymorphic_on": "type",
        "inherit_condition": (id == Group.id)
    }


class GrandchildGroup(ChildGroup):
    id = Column(ForeignKey("childgroup.id"), primary_key=True, nullable=False)
    __mapper_args__ = {
        "polymorphic_identity": "grandchildgroup",
        "polymorphic_on": "type",
        "inherit_condition": (id == ChildGroup.id)
    }
```

1.3 Version of Code
```
db_url = str(sa.engine.url.URL(
    drivername="postgresql",
    username="",
    password="",
    host="localhost",
    port=5432,
    database="test",
))

engine = create_engine(db_url, echo=True)
ORMBase.metadata.create_all(engine)
session = sessionmaker(bind=engine)()

inspect(GrandchildGroup).c
```

1.4 Version of Code
```
db_url = sa.engine.url.URL.create(
        drivername="postgresql",
        username="",
        password="",
        host="localhost",
        port=5432,
        database="test",
    )

engine = create_engine(db_url, echo=True)
ORMBase.metadata.create_all(engine)

with Session(engine) as session:
     inspect(GrandchildGroup).c
```

When we use inspect to get the Mapper object and look at its column.

In 1.3 we see 3 columns
![image](https://github.com/sqlalchemy/sqlalchemy/assets/29122286/fa32d5a2-558d-46e2-ac75-608e6677151a)

But in 1.4, we see 4 columns (id column is duplicated)
![image](https://github.com/sqlalchemy/sqlalchemy/assets/29122286/5eb29a96-100e-4dab-bdc3-f0664c7913c9)

Is that expected?

**Comment by zzzeek** [comment]
Probably not but I would have to see what ChildGroup reports, mostly to see if this is idiosyncratic towards more than two levels of inheritance or not, and also would really only be interested in what 2.0 reports at this point since 1.4 is in maintenance...there may have been some 3-level inheritance issues fixed in 2.0.

**Reply by liverpool1026** [reply]
I tested with 2.0, it is showing the same result as 1.4 is.

![image](https://github.com/sqlalchemy/sqlalchemy/assets/29122286/c94c8d37-8603-4271-a442-fc87efed0846)

Not sure what you mean `but I would have to see what ChildGroup reports`

**Reply by zzzeek** [reply]
it means is the behavior consistent for any inherited class, or only ones that are more than one level deep.  it's consistent:

```py
print(inspect(Group).c)
print(inspect(ChildGroup).c)
print(inspect(GrandchildGroup).c)

```

```
ColumnCollection(group.type, group.id, group.group_name)
ColumnCollection(group.type, group.id, group.group_name, childgroup.id)
ColumnCollection(group.type, childgroup.id, group.group_name, grandchildgroup.id)

```

the `mapper.c` attribute is a very old legacy attribute that is not really useful in modern SQLAlchemy so there should be no reason to rely upon it for anything

**Reply by zzzeek** [reply]
columncollection can accommodate duplicate columns as of 1.4, when a mapper takes in column properties it also adapts those from the immediate inheriting mapper.  so in this case "id" gets added a second time for the superclass.

**Reply by liverpool1026** [reply]
That makes sense,

In our case, we are looking to get the set of columns (so no duplicates of id).

we found `mapper.column_attrs` gives us the correct set, so is it a good option instead of `mapper.c`
